Do Not Leave Logging on In Production Cherrypy Installations!
#time 05-05-07 15:19
#keywords cherrypy, python
<p>On March 10, I tried to roll out a new version of my <a 
href="http://billmill.org:9561/cherry_blossom"> web site software</a>. However, 
after about 20 minutes of normal operation, the website would suddenly start 
throwing <a 
href="http://groups.google.com/group/cherrypy-users/msg/33fe863cd647ce21">strange 
tracebacks</a>, and returning them with x-php-fastcgi mime types, which is 
weird.<p>

For the past 2 months, I've been trying to figure out the cause of this error 
in the little time I have to work on my website, but it proved extremely 
frustrating. In the course of this work, I rewrote all of my templates from <a 
href="http://www.cheetahtemplate.org/">cheetah</a> to <a 
href="http://makotemplates.org">mako</a>, upgraded to python 2.5, and tried 
heroically to reproduce the problem.<p>
To make a long, frustrating story short, cherrypy 2.2 used print statements to 
output log messages. If that failed, no problems occurred. Cherrypy 3.0, on the 
other hand, uses the python logging module, and if stdout fails, it tries 
stderr. If <i>that</i> fails, then it throws a "Broken Pipe" error with a long 
traceback. Thanks to <a 
href="http://groups.google.com/group/cherrypy-users/msg/98ba0410443446cf">Robert 
Brewer</a> for the inside info on what was happening.<p>
So, if you're getting "IOError: [Errno 32] Broken pipe", try setting 
log.screen=False in the [global] section of your configuration. It might save 
you a lot of trouble.<p>
